### PR TITLE
Gtk bridge

### PR DESCRIPTION
**Pull request by skef**
This isn't really a request -- just a context for discussion. 

This is my current gtk bridge code. Configure with `--enable-gdk --enable-gtk-bridge` . I haven't verified a lot of things, including whether the gxdraw version still compiles correctly. I also don't know what this will do under, for example, Windows. 

After building throw a file with this content in your `.config/fontforge/python` directory:
```
import fontforge

fontforge.registerMenuItem(lambda x,y: y.addWindow(), None, None, "Font", None, "Add Window")
```

Then run `Tools -> Add Window` from some font window. 

Note that:

1. Cross-modality works
2. The created windows go into the same window manager slot (under X). 
3. The icon decorations aren't quite right, but close. 
4. And most importantly inspect `gdraw/gtkbridge.c:make_window` and note that it isn't doing anything "custom". GTK pretty much works "as advertised". 

There are two directions to go with a stable version of this:

1. Gradually throw various windows and dialogs "over the fence". This need not be a process that ever completes (although it sounds like you anticipate GDK4 being a problem). The obvious low-hanging fruit here are the file dialogs. I would actually advocate doing a minimal set with at least those not so much because I think it's needed, but to stave off bit rot. Collab rotted away and anything like this could too without ongoing care. On the other hand, GTK and GDK are sufficiently well-associated that I suspect any bug that came up could be addessed in one way or another*.
2. Allowing python GTK UI extension. That would sort of work with what I have now. However, if we want to support longstanding windows linked to particular fonts or CV windows (and we probably would), we'll have to set up an appropriate signal set. Things like "the window now shows a new character" and possibly even "a layer has been changed". Not a huge deal, though (probably). 

In theory this could also work with GXDraw by "promoting" spare X events to GDK events and passing them up the pipeline. But that seems a really unnecessary avenue for exploration, just asking for hours of investigation and debugging for a combination that doesn't add anything.

### CODE REVIEW

**Comment by skef** [comment]
I just added a text edit box to the window to show that key events are being routed reasonably.

**Comment by skef** [comment]
(This is probably clear, but obviously I haven't done 100 commits on this little project. Pulling FontForge's master into yours may prompt github to list only the relevant changes. Or you can do whatever with my branch.)

**Comment by jtanx** [comment]
There's something weird about your branch, because my master is relatively up to date with master. So just confirming, you're ok if I push to your branch?

**Comment by skef** [comment]
> There's something weird about your branch, because my master is relatively up to date with master. So just confirming, you're ok if I push to your branch?

I am. I just made a copy in another branch to be extra careful and so I can pull the result of whatever you do.

**Comment by jtanx** [comment]
cool, I rebased it to master.

File: fontforgeexe/pythonui.c [hunk 1]
```diff
@@ -596,12 +597,12 @@ static PyObject *PyFFFont_addGtkWindowToMainEventLoop(PyFF_Font *self, PyObject
     if ( !PyArg_ParseTuple( args, "i", &v ))
         return( NULL );
 
-    gpointer gdkwindow = gdk_xid_table_lookup( v );
```
Comment (skef) [inline fontforgeexe/pythonui.c:599]: Oh -- these changes are unrelated to the functionality. They're basically just to get this part of the code to compile. It would take more work to revive the gxdraw collab hooks.

File: gdraw/Makefile.am [hunk 2]
```diff
@@ -41,7 +41,8 @@ libgdraw_la_SOURCES = choosericons.c ctlvalues.c drawboxborder.c	\
 	gdrawable.c gspacer.c xkeysyms_unicode.c colorP.h gdrawP.h	\
 	gimagebmpP.h gresourceP.h gxcdrawP.h fontP.h ggadgetP.h		\
 	gpsdrawP.h gwidgetP.h gxdrawP.h hotkeys.c hotkeys.h		\
-	div_tables.c ggdkdraw.c ggdkcdraw.c ggdkdrawlogger.c ggdkdrawP.h
+	div_tables.c ggdkdraw.c ggdkcdraw.c ggdkdrawlogger.c            \
+	gtkbbr.c gtkbridge.c ggdkdrawP.h
```
Comment (skef) [inline gdraw/Makefile.am:45]: `gtkbbr.c` is a (probably) temporary code context for moving between FontForge and GTK. It's basically an include file issue; if you try to include all of everything it gets ugly. Easier to have a neutral location to sort through pointer chains.

File: gdraw/ggdkdraw.c [hunk 3]
```diff
@@ -407,6 +423,16 @@ static void _GGDKDraw_CenterWindowOnScreen(GGDKWindow gw) {
     gdk_window_move(gw->w, gw->pos.x, gw->pos.y);
 }
 
+static GdkPixbuf *_GGDKDraw_IconToPixbuf(GGDKWindow icon) {
```
Comment (skef) [inline gdraw/ggdkdraw.c:426]: Pulled out so it can be used when setting the default icon.

File: gdraw/ggdkdraw.c [hunk 4]
```diff
@@ -857,6 +878,7 @@ static void _GGDKDraw_DispatchEvent(GdkEvent *event, gpointer data) {
     if (event_time != GDK_CURRENT_TIME) {
         gdisp->last_event_time = event_time;
     }
+    GdkEventCrossing *cross;
```
Comment (skef) [inline gdraw/ggdkdraw.c:881]: This is cruft I forgot to remove -- that approach didn't work.

File: gdraw/ggdkdraw.c [hunk 5]
```diff
@@ -865,6 +887,10 @@ static void _GGDKDraw_DispatchEvent(GdkEvent *event, gpointer data) {
         return;
     } else if ((gw = g_object_get_data(G_OBJECT(w), "GGDKWindow")) == NULL) {
         //Log(LOGDEBUG, "MISSING GW!");
+#ifdef FONTFORGE_CAN_USE_GTK_BRIDGE
+	// If the target window is not GGDKDraw's, pass the event onto GTK
+	gtkb_ProcessEvent(gdisp->gtkb_state, event);
+#endif
```
Comment (skef) [inline gdraw/ggdkdraw.c:893]: This is the surprisingly simple core of the bridge. Find an event destined for a window you don't know about? Instead of throwing it away, send it up the GTK stack.

File: gdraw/ggdkdraw.c [hunk 6]
```diff
@@ -1221,6 +1247,10 @@ static void GGDKDrawSetDefaultIcon(GWindow icon) {
     GGDKWindow gicon = (GGDKWindow)icon;
     if (gicon->is_pixmap) {
         gicon->display->default_icon = gicon;
+#ifdef FONTFORGE_CAN_USE_GTK_BRIDGE
```
Comment (skef) [inline gdraw/ggdkdraw.c:1250]: Grab an icon to use for window decoration. Sort of works now but the image comes out funny on my (KDE Plasma) system.

File: gdraw/ggdkdraw.c [hunk 7]
```diff
@@ -2557,6 +2593,17 @@ G_GNUC_END_IGNORE_DEPRECATIONS
...
+    // XXX These may be preferable to the settings in
+    // _GGDKDraw_CreateWindow, but using both should be OK
+    // as long as the same strings are used
+    gdk_set_program_class(GResourceProgramName);
+    // GTK derives the WM_CLASS from this
+    g_set_prgname(GResourceProgramName);
+
+    gdisp->gtkb_state = gtkb_CreateState();
+#endif
```
Comment (skef) [inline gdraw/ggdkdraw.c:2605]: This makes icons and task bar elements in X group with FontForge elements. I put these here because they're gobject and gdk level rather than strictly gtk.

**Comment by khaledhosny** [comment]
There is an idea I wanted to do for some time (but apparently I lack the skill to do it); make FontForge a pseudo GTK app which means all windows are GTK windows but instead of using GTK widgets, window content is just a giant custom GTK widget (which is is basically a Cairo context that FontForge draws whatever it wants on), and then incrementally convert dialogs and windows one by one to use GTK widgets. Not sure if the work proposed here can help in this direction.

**Comment by jtanx** [comment]
Yeah, I floated that idea too (I think I left a comment on #3506), one thing that will need to be fixed before that can happen is restricting all drawing calls to only occur during expose events. The other question is whether or not gtk exposes enough to allow us to do some of the things that gdraw requires (or I guess the alternative is to adjust gdraw enough to fit)

**Comment by skef** [comment]
The primary role this model could play in that context is letting the development stretch out over time. It's (probably) already at the point where you could open a GTK modal dialog "conceptually transient" for a GGDKDraw window. That means you can start moving the dialog code over. 

For something close to @khaledhosny's model you could pick an individual window like the FontView, reimplement the usual "outside" stuff in standard GTK widgets, and then (when necessary) port the "guts" to a hybrid GTK/GXDraw widget. I suspect that's an easier project than trying to get all of GDraw crammed into GTK at once.
